src/concepts/mods.md

```diff
@@ -1 +1,506 @@
+what is in the signature of a structure, and nothing more. This is not strictly
+true - if we have a transparently ascribed structure like `RectArray`, then the
+user of the `RectArray` structure can treat the `'a RectArray.array` type as
+synonymous with `'a list list`, and thus be privy to the "internals" of the
+structure, even though it is not explicitly defined within the structure.
+
+We will now compare this to if we had instead used opaque ascription. Consider
+the following two implementations of the `ARRAY` signature:
+
+```sml
+structure ListArray :> ARRAY =
+struct
+    type 'a array = 'a list * int
+
+    fun gen 0 v = []
+      | gen n v = v :: gen (n - 1) v
+
+    fun new (height, width) v =
+        (gen (height * width) v, width)
+
+    fun get (A, w) (y, x) = List.nth (A, w * y + x)
+
+    fun set (A, w) (y, x) v = List.update (A, w * y + x, v)
+end
+
+structure RectArray :> ARRAY =
+struct
+    type 'a array = 'a list list
+
+    fun row 0 v = []
+      | row n v = v :: row (n - 1) v
+
+    fun new (0, width) v = []
+      | new (height, width) v =
+        row w v :: new (height - 1, width) v
+
+    fun get A (y, x) =
+        List.nth (List.nth (A, y), x)
+
+    fun set A (y, x) v =
+        let
+          val row = List.nth (A, y)
+        in
+          List.update (A, y, List.update (row, x, v))
+        end
+end
+```
+
+Note that the latter implementation is exactly the same as the previous
+`RectArray` structure, with the exception that is now _opaquely ascribed_. This
+is from using `:>` between the structure and signature names, which signals
+opaque ascription rather than transparent ascription.
+
+> **[Opaque ascription]**: Ascribing a structure to a signature where all
+> abstract types are hidden to the user of the structure.
+
+By "hidden", we mean precisely in the same way that the abstract types are
+"visible" in transparent ascription. Abstract types are truly that, _abstract_,
+and thus the user of either the `RectArray` or `ListArray` structure are
+incapable of knowing precisely what either the `'a ListArray.array` or `'a
+RectArray.array` types are. The only way to construct a value of type `'a
+array`, from either structure, is by using either respective `new` function, and
+similarly with manipulating an existing array.
+
+Why would we want to have opaque ascription? This is precisely for the reason
+that we have been motivating for the entire chapter, which is abstraction. A
+fundamental idea of having a powerful type system is to _make illegal states
+unrepresentable_. Instead of having a precondition that must be obeyed in a
+given function, we instead would like it so that the type system enforces these
+preconditions for us instead - similarly to how for a function of type `int ->
+int` with precondition "TRUE", if we were to instead write it in a dynamically
+typed ("typeless") language, we may instead say that it has a precondition that
```

Comment:
this analogy is a little strange imo. I really like the example you provide later about a "bad" 2D array and think that is self explanatory